### ISSUE TITLE
Complicated vector dimensions

### ISSUE THREAD

**Issue by wcdawn** [mfem/mfem#2055]
I am solving a coupled system of PDEs. They are diffusion-like (second order form of the steady-state transport equation). Specifically, they form an eigenvalue problem that I am solving with SLEPc.

I would like to build two bilinear forms for the generalized eigenvalue problem (Ax =\lambda Mx).
To do this, I have problem using the "vector dimension" in MFEM. For these problems, `vdim` is on the order 100 (very small) to 100,000 (very large).
So far, I have implemented a new wrapper class inherited from `mfem::BilinearFormIntegrator` so I don't have to rewrite all of the integrators. The main effect is what follows.

```cpp
void VectorIntegratorWrapper::AssembleElementMatrix(const FiniteElement& el,
                                                    ElementTransformation& Trans,
                                                    DenseMatrix& elmat)
{

  bfi[0]->AssembleElementMatrix(el, Trans, pelmat);
  const int dof{pelmat.Width()};

  elmat.SetSize(vdim*dof);
  elmat = 0.0;
  elmat.AddMatrix(pelmat, row_idx[0], col_idx[0]);

  for (int i{1}; i < bfi.Size(); ++i)
  {
    bfi[i]->AssembleElementMatrix(el, Trans, pelmat);
    elmat.AddMatrix(pelmat, row_idx[i]*dof, col_idx[i]*dof);
  }

  return;

}
```

While this did save me from reimplementing the individual integrators that I need (as I did in #1727), writing the code to build all of the `BilinearFormIntegrator` objects is tedious and uses a non-trivial amount of memory (10-40 GB). Also, this method requires evaluating the same integration point many times which is not ideal.

I'd like to implement something like I did earlier to optimize this process but I have a few questions.
1. Is there a better method for going about this? I'll need to implement vector support for `DGTraceIntegrator`, `ConvectionIntegrator`, and make some additions to `VectorDiffusionIntegrator`.
2. Some of these integrators naturally use `VectorCoefficient` and `MatrixCoefficient` so my earlier method will not work. For example, in my implementation of `ConvectionIntegrator`, I would need a two-dimension array of `Vector` (or `VectorCoefficient`). I also need a two-dimension array of `Matrix` (or `MatrixCoefficient`) for my purposes with `DiffusionIntegrator`. What container should I use for this? `Array2D` sounds like a good idea but this will not be POD. Maybe an `Array<Vector*>` with `row_idx` and `col_idx`?
3. Would it be better to introduce a new `Coefficient` for this? Something like `TensorCoefficient`?

Thanks for the help!

**Comment by wcdawn** [comment]
I've implemented some tensor classes on the branch `wcdawn/TensorCoefficient`. It works and shows the idea that I was trying to communicate but it could use some cleanup.

These routines still require a lot of memory. Partially because the dimension of these problems is so large and partially because `BilinearFormIntegrator::AssembleElementMatrix` will always return a dense matrix.

P.S. I'm looking forward to running these problems with the new PETSc+CUDA capabilities! :)

**Comment by jandrej** [comment]
Have you considered, although solving a linear problem, using the BlockNonlinearForm for this? You can iterate over the list of fields inside the integrator and couple them inside _one_ integrator.

**Comment by wcdawn** [comment]
@jandrej I did something similar using `BlockOperator` and filling it with `PetscParMatrix` since I'm sending everything to SLEPc anyways. To solve the same problem, it used more memory. Also, with `vdim` on the order of 10,000 it is important to order `byVDIM` to reduce bandwidth or else memory is all over the place and that is not possible with the `BlockOperator` or `BlockNonlinearForm`.

If I used `BlockNonlinearForm` wouldn't I have to write all my own integrators?
